File: Extensions/Tessa.Extensions.Client/UI/PnrServiceNoteUIExtension.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows;
using Tessa.Cards;
using Tessa.Extensions.Shared.PnrCards;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Controls;

namespace Tessa.Extensions.Client.UI
{
    public sealed class PnrServiceNoteUIExtension : CardUIExtension
    {
        private readonly ICardRepository cardRepository;
        public PnrServiceNoteUIExtension(ICardRepository cardRepository)
        {
            this.cardRepository = cardRepository;
        }

        //private async void SetLegalEntityIndex(Guid organizationID, ICardModel cardModel)
        //{
        //    CardRequest request = new CardRequest
        //    {
        //        RequestType = Shared.PnrRequestTypes.GetIndexLegalEntityRequestTypeID,
        //        Info =
        //        {
        //            { "organizationID", organizationID }
        //        }
        //    };

        //    CardResponse response = await cardRepository.RequestAsync(request);

        //    Tessa.Platform.Validation.ValidationResult result = response.ValidationResult.Build();
        //    TessaDialog.ShowNotEmpty(result);
        //    if (result.IsSuccessful)
        //    {
        //        ///cardModel.Card.Sections["PnrServiceNote"].Fields["LegalEntityIndexID"] = response.Info.Get<Guid?>("LegalEntityID"); ;
        //        cardModel.Card.Sections["PnrServiceNote"].Fields["LegalEntityIndexIdx"] = response.Info.Get<string>("LegalEntityIdx");
        //    }
        //}

        // установка видимости блоков, зависимых от Типа служебной записки
        private void SetNTDependenceVisibility(Guid? serviceNoteTypeID, ICardModel cardModel)
        {
            IBlockViewModel personnelBlock = cardModel.Blocks["PersonnelBlock"];
            personnelBlock.BlockVisibility = serviceNoteTypeID == PnrServiceNoteTypes.Personnel ? Visibility.Visible : Visibility.Collapsed;

            IBlockViewModel conclusionContractsBlock = cardModel.Blocks["ConclusionContractsBlock"];
            conclusionContractsBlock.BlockVisibility = serviceNoteTypeID == PnrServiceNoteTypes.ConclusionContracts ? Visibility.Visible : Visibility.Collapsed;

            IBlockViewModel conclusionContractsDUPBlock = cardModel.Blocks["ConclusionContractsDUPBlock"];
            conclusionContractsDUPBlock.BlockVisibility = serviceNoteTypeID == PnrServiceNoteTypes.ConclusionContractsDUP ? Visibility.Visible : Visibility.Collapsed;

            IBlockViewModel financialActivitiesBlock = cardModel.Blocks["FinancialActivitiesBlock"];
            financialActivitiesBlock.BlockVisibility = serviceNoteTypeID == PnrServiceNoteTypes.FinancialActivities ? Visibility.Visible : Visibility.Collapsed;

            IBlockViewModel workQuestionsBlock = cardModel.Blocks["WorkQuestionsBlock"];
            workQuestionsBlock.BlockVisibility = serviceNoteTypeID == PnrServiceNoteTypes.WorkQuestions ? Visibility.Visible : Visibility.Collapsed;

            workQuestionsBlock.Rearrange();
        }

        private void ChangeControlsActiveValidation(Guid? serviceNoteTypeID, ICardModel cardModel)
        {
            if (cardModel.Controls.TryGet("ConclusionContractsTheme", out IControlViewModel ContractsThemeControl) &&
                cardModel.Controls.TryGet("FinancialActivitiesTheme", out IControlViewModel ActivitiesThemeControl))
            {
                ContractsThemeControl.IsRequired = serviceNoteTypeID == PnrServiceNoteTypes.ConclusionContracts;
                ActivitiesThemeControl.IsRequired = serviceNoteTypeID == PnrServiceNoteTypes.FinancialActivities;
                ContractsThemeControl.NotifyUpdateValidation();
                ActivitiesThemeControl.NotifyUpdateValidation();
            }
            
        }

        private async void SetUserDepartmentInfo(Guid? authorID, ICardModel cardModel, Boolean isInitiator)
        {
            CardRequest request = new CardRequest
            {
                RequestType = Shared.PnrRequestTypes.GetUserDepartmentInfoRequestTypeID,
                Info =
                {
                    { "authorID", authorID }
                }
            };

            CardResponse response = await cardRepository.RequestAsync(request);

            Tessa.Platform.Validation.ValidationResult result = response.ValidationResult.Build();
            TessaDialog.ShowNotEmpty(result);
            if (result.IsSuccessful)
            {
                if (isInitiator)
                {
                    cardModel.Card.Sections["PnrServiceNote"].Fields["DepartmentID"] = response.Info.Get<Guid?>("DepartmentID");
                    cardModel.Card.Sections["PnrServiceNote"].Fields["DepartmentName"] = response.Info.Get<string>("Name");
                    cardModel.Card.Sections["PnrServiceNote"].Fields["DepartmentIdx"] = response.Info.Get<string>("Index");
                }
                else
                {
                    cardModel.Card.Sections["PnrServiceNote"].Fields["DestinationDepartmentID"] = response.Info.Get<Guid?>("DepartmentID");
                    cardModel.Card.Sections["PnrServiceNote"].Fields["DestinationDepartmentName"] = response.Info.Get<string>("Name");
                    cardModel.Card.Sections["PnrServiceNote"].Fields["DestinationDepartmentIdx"] = response.Info.Get<string>("Index");
                }
            }
        }

        public override Task Initialized(ICardUIExtensionContext context)
        {
            Card card = context.Card;
            ICardModel cardModel = context.Model;

            
            if (card.StoreMode == CardStoreMode.Insert)
            {
                // Заполнить поле Дата регистрации текущей датой
                if (cardModel.Controls.TryGet("RegistrationDate", out IControlViewModel regDateControl))
                {
                    ((DateTimeViewModel)regDateControl).SelectedDate = DateTime.Now;
                }
                // начальная инициализация поля Подразделение по автору
                SetUserDepartmentInfo((Guid?)card.Sections["DocumentCommonInfo"].Fields["AuthorID"], cardModel, true);
            } 
            else
            {
                // Открытие карточки. Есть Тип служебной записки - устанавливаем видимость блоков в зависимости от Типа
                SetNTDependenceVisibility((Guid?)card.Sections["PnrServiceNote"].Fields["ServiceNoteTypeID"], cardModel);
            }

            context.Card.Sections["PnrServiceNote"].FieldChanged += (s, e) =>
            {
                // Тип служебной записки: Visibility блоков
                if (e.FieldName == "ServiceNoteTypeID" && e.FieldValue != null)
                {
                    SetNTDependenceVisibility((Guid)e.FieldValue, cardModel);
                    ChangeControlsActiveValidation((Guid)e.FieldValue, cardModel);

                }

                // Организация ГК Пионер: заполнение Индекс ЮЛ (но он далее нигде не используется)
                //if (e.FieldName == "OrganizationID" && e.FieldValue != null)
                //{
                //    SetLegalEntityIndex((Guid)e.FieldValue, cardModel);
                //}

                // Адресат: подчитываем его подразделение
                if (e.FieldName == "DestinationID" && e.FieldValue != null)
                {
                    SetUserDepartmentInfo((Guid?)e.FieldValue, cardModel, false);
                }
            };

            return base.Initialized(context);
        }
    }
}

File: WebClient SDK/src/PnrScripts/UI/PnrServiceNoteUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { Guid, createTypedField, DotNetType, Visibility } from 'tessa/platform';
import { CardStoreMode } from 'tessa/cards';
import { PnrCardTypes } from '../Shared/PnrCardTypes';
import { PnrRequestTypes } from '../Shared/PnrRequestTypes';
import { PnrServiceNoteTypes } from '../Shared/PnrServiceNoteTypes';
import { CardRequest, CardService } from 'tessa/cards/service';
import { DateTimeViewModel } from 'tessa/ui/cards/controls';
import Moment from 'moment';

export class PnrServiceNoteUIExtension extends CardUIExtension {
  public async initialized(context: ICardUIExtensionContext) {
    if (!Guid.equals(context.card.typeId, PnrCardTypes.PnrServiceNoteTypeID)) {
      return;
    }
    let PnrServiceNote = context.card.sections.tryGet("PnrServiceNote");
    if (!PnrServiceNote) return;
    // Создание карточки
    if (context.card.storeMode === CardStoreMode.Insert)
    {
        const regDateControl = context.model.controls.get('RegistrationDate') as DateTimeViewModel;
        if (!regDateControl) return;
        // Заполнить поле Дата проекта текущей датой
        regDateControl.selectedDate = Moment(Date.now());

      let dci = context.card.sections.tryGet("DocumentCommonInfo");
      if (dci != null) {
        // начальная инициализация поля Подразделение по автору
        await this.SetUserDepartmentInfo(dci.fields.get('AuthorID'), context, true);
      }
    }
    else {
      // Открытие карточки. Есть Тип служебной записки - устанавливаем видимость блоков в зависимости от Типа
      this.SetNTDependenceVisibility(PnrServiceNote.fields.get('ServiceNoteTypeID'), context);
    }

    PnrServiceNote.fields.fieldChanged.add(async (e) => {
      // Тип служебной записки: Visibility блоков
      if (e.fieldName === "ServiceNoteTypeID" && e.fieldValue != null)
      {
         this.SetNTDependenceVisibility(e.fieldValue, context);
         this.ChangeControlsActiveValidation(e.fieldValue, context);
      }
      // Организация ГК Пионер: заполнение Индекс ЮЛ (но он далее нигде не используется)
      // if (e.fieldName == "OrganizationID" && e.fieldValue != null)
      // {
      //   await this.SetLegalEntityIndex(e.fieldValue, context);
      // }
      // Адресат: подчитываем его подразделение
      if (e.fieldName === "DestinationID" && e.fieldValue != null)
      {
        await this.SetUserDepartmentInfo(e.fieldValue, context, false);
      }
    });
  }

  // private async SetLegalEntityIndex(organizationID, context: ICardUIExtensionContext)
  // {
  //   // request к организации получить индекс ЮЛ
  //   let LegalEntityIdx;
  //   let request = new CardRequest();
  //       request.requestType = PnrRequestTypes.GetIndexLegalEntityRequestTypeID;
  //       request.info["organizationID"] = createTypedField(organizationID, DotNetType.Guid);
  //   // проверка ответа
  //   let response = await CardService.instance.request(request);
  //   if (response.validationResult.isSuccessful) {
  //       LegalEntityIdx = response.info.LegalEntityIdx;
  //   }
  //   let PnrServiceNote = context.card.sections.tryGet("PnrServiceNote");
  //   if (LegalEntityIdx != null && PnrServiceNote != null) {
  //     PnrServiceNote.fields.set('LegalEntityIndexIdx', createTypedField(LegalEntityIdx.$value, DotNetType.String))
  //   }
  // }

  /** установка видимости блоков, зависимых от Типа служебной записки */
  private SetNTDependenceVisibility(serviceNoteTypeID, context: ICardUIExtensionContext)
  {
    const personnelBlock = context.model.blocks.get('PersonnelBlock');
    if (personnelBlock) {
      personnelBlock.blockVisibility = serviceNoteTypeID === PnrServiceNoteTypes.Personnel ? Visibility.Visible : Visibility.Collapsed;
    }
    const conclusionContractsBlock = context.model.blocks.get('ConclusionContractsBlock');
    if (conclusionContractsBlock) {
      conclusionContractsBlock.blockVisibility = serviceNoteTypeID === PnrServiceNoteTypes.ConclusionContracts ? Visibility.Visible : Visibility.Collapsed;
    }
    const conclusionContractsDUPBlock = context.model.blocks.get('ConclusionContractsDUPBlock');
    if (conclusionContractsDUPBlock) {
      conclusionContractsDUPBlock.blockVisibility = serviceNoteTypeID === PnrServiceNoteTypes.ConclusionContractsDUP ? Visibility.Visible : Visibility.Collapsed;
    }
    const financialActivitiesBlock = context.model.blocks.get('FinancialActivitiesBlock');
    if (financialActivitiesBlock) {
      financialActivitiesBlock.blockVisibility = serviceNoteTypeID === PnrServiceNoteTypes.FinancialActivities ? Visibility.Visible : Visibility.Collapsed;
    }
    const workQuestionsBlock = context.model.blocks.get('WorkQuestionsBlock');
    if (workQuestionsBlock) {
      workQuestionsBlock.blockVisibility = serviceNoteTypeID === PnrServiceNoteTypes.WorkQuestions ? Visibility.Visible : Visibility.Collapsed;
    }
  }

  private ChangeControlsActiveValidation(serviceNoteTypeID, context: ICardUIExtensionContext)
        {
          let contractsThemeControl = context.model.controls.get("ConclusionContractsTheme");
          let activitiesThemeControl = context.model.controls.get("FinancialActivitiesTheme");
            if (contractsThemeControl && activitiesThemeControl)
            {
              contractsThemeControl.isRequired = serviceNoteTypeID === PnrServiceNoteTypes.ConclusionContracts;
              activitiesThemeControl.isRequired = serviceNoteTypeID === PnrServiceNoteTypes.FinancialActivities;
              contractsThemeControl.notifyUpdateValidation();
              activitiesThemeControl.notifyUpdateValidation();
            }
        }

  private async SetUserDepartmentInfo(authorID, context: ICardUIExtensionContext, isInitiator: boolean)
  {
    let DepartmentID;
    let DepartmentName;
    let DepartmentIdx;
    let request = new CardRequest();
        request.requestType = PnrRequestTypes.GetUserDepartmentInfoRequestTypeID;
        request.info["authorID"] = createTypedField(authorID, DotNetType.Guid);

    // проверка ответа
    let response = await CardService.instance.request(request);
    if (response.validationResult.isSuccessful) {
      DepartmentID = response.info.DepartmentID;
      DepartmentName = response.info.Name;
      DepartmentIdx = response.info.Index;
    }

    let PnrServiceNote = context.card.sections.tryGet("PnrServiceNote");
    if (DepartmentID != null && DepartmentName != null && DepartmentIdx != null && PnrServiceNote != null) {
      if (isInitiator) {
        PnrServiceNote.fields.set('DepartmentID', createTypedField(DepartmentID.$value, DotNetType.Guid));
        PnrServiceNote.fields.set('DepartmentName', createTypedField(DepartmentName.$value, DotNetType.String));
        PnrServiceNote.fields.set('DepartmentIdx', createTypedField(DepartmentIdx.$value, DotNetType.String));
      }
      else {
        PnrServiceNote.fields.set('DestinationDepartmentID', createTypedField(DepartmentID.$value, DotNetType.Guid));
        PnrServiceNote.fields.set('DestinationDepartmentName', createTypedField(DepartmentName.$value, DotNetType.String));
        PnrServiceNote.fields.set('DestinationDepartmentIdx', createTypedField(DepartmentIdx.$value, DotNetType.String));
      }
    }
  }
}